File: JavaExercise/src/top/interview/ClimbStairs.java
package top.interview;

/***
 * https://leetcode.com/explore/featured/card/top-interview-questions-easy/97/dynamic-programming/569/
 *
 * 画图理解、、、列方程
 */
public class ClimbStairs {
    public ClimbStairs(){

    }
    public int climbStairs(int n) {
        if (n == 0 || n == 1 || n == 2) {
            return n;
        }else {
            int[] steps = new int[n];
            steps[0] = 1;
            steps[1] = 2;
            for (int i = 2; i < n; i++){
                steps[i] = steps[i-1] + steps[i-2];
            }
            return steps[n-1];
        }
    }

    public static void main(String[] args) {
        System.out.println(new ClimbStairs().climbStairs(3));
    }
}

File: tencent/DynamicProgramming/climbStairs.py
# -*- coding: utf-8 -*-
"""

   File Name：     climbStairs
   Author :       jing
   Date：          2020/3/20

   https://leetcode-cn.com/explore/interview/card/tencent/226/dynamic-programming/921/

   爬楼梯
"""


class Solution:
    def climbStairs(self, n: int) -> int:
        if n <= 2:
            return n
        result = []
        i = 0
        while i <= 2:
            result.append(i)
            i += 1

        for i in range(3, n+1):
            result.append(result[i-2] + result[i-1])   # 状态转移方程： 要么跨一步到n,要么2步，所以是result[i-2] + result[i-1]

        return result[n]


if __name__ == '__main__':
    print(Solution().climbStairs(4))



